File: Heaps/LastStoneWeight.java
package Heaps;

import java.util.Comparator;
import java.util.PriorityQueue;

class LastStoneWeight {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder());

        for (int stone : stones)
            maxHeap.add(stone);

        while (maxHeap.size() > 1) {
            int heaviest = maxHeap.poll();
            int secondHeaviest = maxHeap.poll();
            int remainingStone = heaviest - secondHeaviest;

            if (remainingStone > 0)
                maxHeap.add(remainingStone);
        }

        return maxHeap.isEmpty() ? 0 : maxHeap.peek();
    }
}

// Time Complexity: O(n log n) - Building the max heap takes O(n) and each of
// the n-1 iterations involves a poll and possibly an add operation,
// each of which takes O(log n).
// Space Complexity: O(n) - In the worst case, all stones could be added to the
// max heap, resulting in O(n) space usage.

File: Heaps/last_stone_weight.py
from heapq import heapify_max, heappop, heappop_max, heappush, heappush_max
from typing import List


class LastStoneWeight:
    def lastStoneWeight(self, stones: List[int]) -> int:
        max_heap = [stone for stone in stones]
        heapify_max(max_heap)
        while len(max_heap) >= 2:
            heaviest = heappop_max(max_heap)
            second_heaviest = heappop_max(max_heap)
            if heaviest != second_heaviest:
                new_stone = heaviest - second_heaviest
                heappush_max(max_heap, new_stone)

        return max_heap[0] if max_heap else 0

# Time complexity: O(n log n) where n is the number of stones. 
# This is because we need to build a max heap from the list of stones, which takes O(n) time, 
# and then we may need to perform up to n-1 operations of removing the two heaviest stones 
# and adding a new stone back to the heap, each of which takes O(log n) time.
# Space complexity: O(n) for the max heap, where n is the number of stones.